Migrate 'torch.dot' from TH to Aten (CUDA)

huangyi1979 opened the pull request:
Summary:
Support **float16** / **float32** / **float64**.
Avoid creating output result on CPU.

Both of two tensors must be on GPU.

ngimel commented:
Fixes #24552. CI errors are real, likely because you don't restore cublas pointer mode.

kshitij12345 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -432,6 +431,36 @@ void ger(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar
+template <>
+void dot<at::Half>(CUDABLAS_DOT_ARGTYPES(at::Half)) {
+  cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+  TORCH_CUDABLAS_CHECK(cublasDotEx(
```
Hi,

https://github.com/pytorch/pytorch/blob/502ec8f7f76619e6cbb1efb0e23eb8c5438daed2/aten/src/THC/THCBlas.cu#L80-L94

ROCM case is missing for Half

kshitij12345 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -432,6 +431,52 @@ void ger(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar
+
+template <>
+void dot<float>(CUDABLAS_DOT_ARGTYPES(float)) {
+  TORCH_CUDABLAS_CHECK(cublasSdot(handle, n, x, incx, y, incy, result));
```
```
  if (n == 1) {
    incx = 1;
    incy = 1;
  }
```
I guess that is missing.

kshitij12345 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -432,6 +431,52 @@ void ger(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar
+
+template <>
+void dot<double>(CUDABLAS_DOT_ARGTYPES(double)) {
+  TORCH_CUDABLAS_CHECK(cublasDdot(handle, n, x, incx, y, incy, result));
```
```
  if (n == 1) {
    incx = 1;
    incy = 1;
  }
```
I guess that is missing.

kshitij12345 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -432,6 +431,52 @@ void ger(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar
+  TORCH_CUDABLAS_CHECK(rocblas_hdot(
+      handle,
+      n,
+      reinterpret_cast<rocblas_half*>(x),
```
Based on the CI error and argument types. This should probably be `const_cast<rocblas_half*>(reinterpret_cast<const rocblas_half*>(x))`

ngimel commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -337,11 +337,81 @@ Tensor& addbmm__cuda(Tensor& self,
+    incy = 1;
+  }
+  TORCH_CHECK(
+      (n <= INT_MAX) && (incx <= INT_MAX) && (incy <= INT_MAX),
```
you have to first check that self.numel(), self.stride(0) and other.stride(0) (which are int64_t) fit in int type, and then convert. This check is guaranteed to pass, because n, incx and incy are int already.

huangyi1979 replied:
oops, good catch

ngimel commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -462,6 +461,47 @@ void ger(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar
+      CUDA_R_16F,
+      CUDA_R_32F));
+#elif HIP_VERSION >= 210
+  cublasSetStream(handle, at::cuda::getCurrentCUDAStream().stream());
```
you don't need it, stream is set in a call to getCurrentCUDABlasHandle.

Commit:
Migrate 'torch.dot' from TH to Aten (CUDA) (#40646)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40646

Support double, float, at::Half.
Avoid creating output result on CPU.

Both of two tensors must be on GPU.

Reviewed By: ngimel

Differential Revision: D22258840

fbshipit-source-id: 27934583d4415711305a11f2be6bef53ce3b38f0
